gui/wxpython/main_window/notebook.py

```diff
@@ -0,0 +1,132 @@
+    def __init__(self, parent, mapdisplay, size, title):
+        wx.Frame.__init__(self, parent=parent, title=title)
+        self.mapdisplay = mapdisplay
+        self.SetSize(size)
+        self.sizer = wx.BoxSizer(wx.HORIZONTAL)
+        self.SetSizerAndFit(self.sizer)
+
+        self.Bind(wx.EVT_CLOSE, self.OnClose)
+
+    def closeFrameNoEvent(self):
+        """Close frame without generating OnClose event."""
+        self.Unbind(wx.EVT_CLOSE)
+        self.Close()
+        self.Bind(wx.EVT_CLOSE, self.OnClose)
+
+    def OnClose(self, event):
+        """Close frame and associated layer notebook page."""
+        self.mapdisplay.OnCloseWindow(event=None, askIfSaveWorkspace=True)
+
+
+class MapNotebook(aui.AuiNotebook):
+    """Map notebook class. Overrides some AuiNotebook classes.
+    Takes into consideration the dock/undock functionality.
+    """
+
+    def __init__(
+        self,
+        parent,
+        agwStyle=aui.AUI_NB_DEFAULT_STYLE | aui.AUI_NB_TAB_EXTERNAL_MOVE | wx.NO_BORDER,
+    ):
+        self.parent = parent
+        super().__init__(parent=self.parent, id=wx.ID_ANY, agwStyle=agwStyle)
+
+        self.SetArtProvider(SimpleTabArt())
+
+        # bindings
+        self.Bind(
+            aui.EVT_AUINOTEBOOK_PAGE_CHANGED,
+            lambda evt: self.GetCurrentPage().onFocus.emit(),
+        )
+        self.Bind(aui.EVT_AUINOTEBOOK_PAGE_CLOSE, self.OnClose)
+
+    def UndockMapDisplay(self, page):
+        """Undock active map display to independent MapFrame object"""
+        index = self.GetPageIndex(page)
+        text = self.GetPageText(index)
+        original_size = page.GetSize()
+        self.RemovePage(index)
+        frame = MapPageFrame(
+            parent=self.parent, mapdisplay=page, size=original_size, title=text
+        )
+        page.Reparent(frame)
+        page.closingFrame.connect(lambda index = index: self.DeletePage(index))
+        page.SetDockingCallback(self.DockMapDisplay)
+        frame.sizer.Add(page, proportion=1, flag=wx.EXPAND)
+        frame.Show()
+        page.Show()
+        page.onFocus.emit()
+
+    def DockMapDisplay(self, page):
+        """Dock independent MapFrame object back to Aui.Notebook"""
+        frame = page.GetParent()
+        page.Reparent(self)
+        page.SetDockingCallback(self.UndockMapDisplay)
+        self.AddPage(page, frame.GetTitle())
+        frame.closeFrameNoEvent()
+
+    def AddPage(self, *args, **kwargs):
+        """Overrides Aui.Notebook AddPage method.
+        Adds page to notebook and makes it current"""
+        super().AddPage(*args, **kwargs)
+        self.SetSelection(self.GetPageCount() - 1)
+
+    def SetSelectionToPage(self, page):
+        """Overrides Aui.Notebook SetSelectionToPage method.
+        Decides whether to set selection to a MapNotebook page
+        or an undocked independent frame"""
+        self.SetSelection(self.GetPageIndex(page))
+
+        if not page.IsDocked():
+            wx.CallLater(500, page.GetParent().Raise)
```

Comment:
This seems to solve the RuntimeError:

```suggestion
            frame = page.GetParent()
            wx.CallLater(500, lambda: frame.Raise if frame else None)
```